### PR TITLE
Vectorize LeadingZeroCount & TrailingZeroCount.

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Using code from #98275 and 1da8d2fb189e3748e3f400dc3d5eeeb310e844c5. Contributes to #97193.

### CODE REVIEW

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.LeadingZeroCount.cs [hunk 1]
```diff
@@ -22,13 +27,56 @@ public static void LeadingZeroCount<T>(ReadOnlySpan<T> x, Span<T> destination)
...
+                }
+
+                if (AdvSimd.IsSupported)
+                {
+                    if (sizeof(T) == 1) return AdvSimd.LeadingZeroCount(x.AsByte()).As<byte, T>();
+                    if (sizeof(T) == 2) return AdvSimd.LeadingZeroCount(x.AsUInt16()).As<ushort, T>();
+                    if (sizeof(T) == 4) return AdvSimd.LeadingZeroCount(x.AsUInt32()).As<uint, T>();
+                }
+
+                throw new NotSupportedException();
```
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.LeadingZeroCount.cs:54]: Based on feedback previously, I stopped doing this and instead asserted, e.g.
```C#
                Debug.Assert(AdvSimd.IsSupported);
                {
                    if (sizeof(T) == 1) return AdvSimd.LeadingZeroCount(x.AsByte()).As<byte, T>();
                    if (sizeof(T) == 2) return AdvSimd.LeadingZeroCount(x.AsUInt16()).As<ushort, T>();
                    if (sizeof(T) == 4) return AdvSimd.LeadingZeroCount(x.AsUInt32()).As<uint, T>();
                }
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.TrailingZeroCount.cs [hunk 2]
```diff
@@ -22,13 +26,52 @@ public static void TrailingZeroCount<T>(ReadOnlySpan<T> x, Span<T> destination)
...
-            public static bool Vectorizable => false; // TODO: Vectorize
+            public static bool Vectorizable =>
+                (AdvSimd.IsSupported && AdvSimd.Arm64.IsSupported && sizeof(T) == 1) ||
+                PopCountOperator<T>.Vectorizable; // http://0x80.pl/notesen/2023-01-31-avx512-bsf.html#trailing-zeros-simplified
+
             public static T Invoke(T x) => T.TrailingZeroCount(x);
-            public static Vector128<T> Invoke(Vector128<T> x) => throw new NotSupportedException();
-            public static Vector256<T> Invoke(Vector256<T> x) => throw new NotSupportedException();
-            public static Vector512<T> Invoke(Vector512<T> x) => throw new NotSupportedException();
+
```
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.TrailingZeroCount.cs:36]: Nit:
```suggestion
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.TrailingZeroCount.cs [hunk 3]
```diff
@@ -22,13 +26,52 @@ public static void TrailingZeroCount<T>(ReadOnlySpan<T> x, Span<T> destination)
...
+                {
+                    return AdvSimd.LeadingZeroCount(AdvSimd.Arm64.ReverseElementBits(x.AsByte())).As<byte, T>();
+                }
+
+                if (PopCountOperator<T>.Vectorizable)
+                {
+                    return PopCountOperator<T>.Invoke(~x & (x - Vector128<T>.One));
+                }
+
+                throw new NotSupportedException();
```
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.TrailingZeroCount.cs:51]: Same comment as on the other file

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.TrailingZeroCount.cs [hunk 4]
```diff
@@ -22,13 +26,52 @@ public static void TrailingZeroCount<T>(ReadOnlySpan<T> x, Span<T> destination)
...
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector256<T> Invoke(Vector256<T> x)
+            {
+                if (PopCountOperator<T>.Vectorizable)
+                {
+                    PopCountOperator<T>.Invoke(~x & (x - Vector256<T>.One));
```
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.TrailingZeroCount.cs:59]: Does this compile?  Seems like it's missing a `return `.  Same below.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.TrailingZeroCount.cs [hunk 5]
```diff
@@ -22,13 +26,52 @@ public static void TrailingZeroCount<T>(ReadOnlySpan<T> x, Span<T> destination)
...
+
+                return Vector256.Create(Invoke(x.GetLower()), Invoke(x.GetUpper()));
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector512<T> Invoke(Vector512<T> x)
+            {
+                if (PopCountOperator<T>.Vectorizable)
+                {
+                    PopCountOperator<T>.Invoke(~x & (x - Vector512<T>.One));
```
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.TrailingZeroCount.cs:70]:
```suggestion
                    return PopCountOperator<T>.Invoke(~x & (x - Vector512<T>.One));
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.TrailingZeroCount.cs (hunk 4 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.TrailingZeroCount.cs:59]: It does. Seems like it was doing extra work for no good reason.

**Review by stephentoub — APPROVED** [review]
Thanks.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.LeadingZeroCount.cs [hunk 6]
```diff
@@ -22,13 +28,56 @@ public static void LeadingZeroCount<T>(ReadOnlySpan<T> x, Span<T> destination)
...
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector128<T> Invoke(Vector128<T> x)
+            {
+                if (Avx512CD.VL.IsSupported)
+                {
+                    if (sizeof(T) == 4) return Avx512CD.VL.LeadingZeroCount(x.AsUInt32()).As<uint, T>();
+                    if (sizeof(T) == 8) return Avx512CD.VL.LeadingZeroCount(x.AsUInt64()).As<ulong, T>();
+                }
+
+                Debug.Assert(AdvSimd.IsSupported);
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.LeadingZeroCount.cs:48]: We want these to be explicit. It makes it work better with the trimmer for other platforms and matches the requirements we have for corelib around using intrinsics (this isn't in corelib, but its still good to follow those practices as its required for special crossgen handling to ever be usable for the assembly).

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.LeadingZeroCount.cs [hunk 7]
```diff
@@ -22,13 +28,56 @@ public static void LeadingZeroCount<T>(ReadOnlySpan<T> x, Span<T> destination)
...
+                {
+                    if (sizeof(T) == 1) return AdvSimd.LeadingZeroCount(x.AsByte()).As<byte, T>();
+                    if (sizeof(T) == 2) return AdvSimd.LeadingZeroCount(x.AsUInt16()).As<ushort, T>();
+                    if (sizeof(T) == 4) return AdvSimd.LeadingZeroCount(x.AsUInt32()).As<uint, T>();
+
+                    throw new NotSupportedException();
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.LeadingZeroCount.cs:54]: Ideally this is a throw helper so the JIT doesn't have to rationalize it isn't hit in the early inlining phases.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.LeadingZeroCount.cs (hunk 6 above)
Comment (stephentoub) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.LeadingZeroCount.cs:48]: The rest of TP is like this, as I'd previously addressed feedback around throws. I suggest this be merged as is, and if we want to change the approach, it be done across all operators en mass together.

**Pull request merged**